utils/s3/client.cc

```diff
@@ -1226,4 +1226,113 @@ future<> client::close() {
+    , _prefix(std::move(prefix))
+    , _max_keys(format("{}", maximum_objects_per_page))
+    , _queue(512 / sizeof(std::optional<directory_entry>))
+{
+}
+
+static std::pair<std::vector<sstring>, sstring> parse_list_of_objects(sstring body) {
+    auto doc = std::make_unique<rapidxml::xml_document<>>();
+    try {
+        doc->parse<0>(body.data());
+    } catch (const rapidxml::parse_error& e) {
+        s3l.warn("cannot parse list-objects-v2 response: {}", e.what());
+        throw std::runtime_error("cannot parse objects list response");
+    }
+
+    std::vector<sstring> names;
+    auto root_node = doc->first_node("ListBucketResult");
+    for (auto contents = root_node->first_node("Contents"); contents; contents = contents->next_sibling()) {
+        auto key = contents->first_node("Key");
+        names.push_back(key->value());
+    }
+
+    sstring continuation_token;
+    auto is_truncated = root_node->first_node("IsTruncated");
+    if (is_truncated && std::string_view(is_truncated->value()) == "true") {
+        auto continuation = root_node->first_node("NextContinuationToken");
+        if (!continuation) {
+            throw std::runtime_error("no continuation token in truncated list of objects");
+        }
+        continuation_token = continuation->value();
+    }
+
+    return {std::move(names), std::move(continuation_token)};
+}
+
+future<> client::bucket_lister::start_listing() {
+    // This is the implementation of paged ListObjectsV2 API call
+    // https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html
+    sstring continuation_token;
+    do {
+        s3l.trace("GET /?list-type=2 (prefix={})", _prefix);
+        auto req = http::request::make("GET", _client->_host, format("/{}", _bucket));
+        req.query_parameters.emplace("list-type", "2");
+        req.query_parameters.emplace("max-keys", _max_keys);
+        if (!continuation_token.empty()) {
+            req.query_parameters.emplace("continuation-token", std::exchange(continuation_token, ""));
+        }
+        if (!_prefix.empty()) {
+            req.query_parameters.emplace("prefix", _prefix);
+        }
+
+        std::vector<sstring> names;
+        try {
+            co_await _client->make_request(std::move(req),
+                [&names, &continuation_token] (const http::reply& reply, input_stream<char>&& in) mutable -> future<> {
+                    auto input = std::move(in);
+                    auto body = co_await util::read_entire_stream_contiguous(input);
+                    auto list = parse_list_of_objects(std::move(body));
+                    names = std::move(list.first);
+                    continuation_token = std::move(list.second);
+                }
+            );
+        } catch (...) {
+            _queue.abort(std::current_exception());
+            co_return;
+        }
+
+        for (auto&& o : names) {
+            co_await _queue.push_eventually(directory_entry{std::move(o)});
```

Comment:
Could you have used generator and co_yield? I guess cleanup is difficult there.

Reply:
seastar's async generator is buggy. a fix is pending on review. see https://github.com/scylladb/seastar/pull/2218

Reply:
Yeah. It's okay to use a queue meanwhile.